netrd/distance/quantum_jsd.py

```diff
@@ -0,0 +1,137 @@
+from .base import BaseDistance
+
+
+class QuantumJSD(BaseDistance):
+    def dist(self, G1, G2, beta=0.1, q=None):
+        """
+        Return the square root of the quantum $q$-Jenson-Shannon divergence
+        between two graphs.
+
+        The generalized Jensen-Shannon divergence compares two graphs by the
+        spectral entropies of their quantum-statistical-mechanical density
+        matrices. It can be written as
+        $$
+        \mathcal{J}_q(\mathbf{\rho} || \mathbf{\sigma}) =
+        S_q\left( \frac{\mathbf{\rho} + \mathbf{\sigma}}{2} \right) -
+        \frac{1}{2} [S_q(\mathbf{\rho}) + S_q(\mathbf{\sigma})],
+        $$
+        where $\mathbf{\rho}$ and $\mathbf{\sigma}$ are density matrices and $q$
+        is the order parameter.
+
+        The density matrix
+        $$
+        \mathbf{\rho} = \frac{e^{-\beta\mathbf{L}}}{Z},
+        $$
+        where
+        $$
+        Z = \sum_{i=1}^{N}e^{-\beta\lambda_i(\mathbf{L})}
+        $$
+        and $\lambda_i(\mathbf{L})$ is the $i$th eigenvalue of the Laplacian
+        matrix $L$, represents an imaginary diffusion process over the network
+        with time parameter $\beta > 0$.
+        
+        For these density matrices and the mixture matrix, we calculate the
+        Rényi entropy of order $q$
+        $$
+        S_q = \frac{1}{1-q} \log_2 \sum_{i=1}^{N}\lambda_i(\mathbf{\rho})^q,
+        $$
+        or, if $q=1$, the Von Neumann entropy
+        $$
+        S_1 = - \sum_{i=1}^{N}\lambda_i(\mathbf{\rho})\log_2\lambda_i(\mathbf{\rho}).
+        $$
+
+        Note that this implementation is not exact because the matrix
+        exponentiation is performed using the Padé approximation.
+
+        Params
+        ------
+
+        G1, G2 (nx.Graph): two networkx graphs to be compared.
+        beta (float): time parameter for diffusion propagator
+        q (float): order parameter for Rényi entropy. If None or 1, use the
+        Von Neumann entropy (i.e., Shannon entropy) instead.
+
+        Returns
+        -------
+
+        dist (float): the distance between G1 and G2.
+
+        """
+
+        if beta <= 0:
+            raise ValueError("beta must be positive.")
+
+        if q and q >= 2:
+            warnings.warn("JSD is only a metric for 0 ≤ q < 2.",
+                          RuntimeWarning)
+
+        def density_matrix(A, beta):
+            """
+            Create the density matrix encoding probabilities for entropies.
+            This is done using a fictive diffusion process with time parameter
+            `beta`.
+            """
+            L = np.diag(np.sum(A, axis=1)) - A
+            rho = expm(-1 * beta * L)
+            rho = rho / np.trace(rho)
+
+            return rho
+
+        def renyi_entropy(X, q=None):
+            """
+            Calculate the Rényi entropy with order `q`, or the Von Neumann
+            entropy if `q` is `None` or 1.
+            """
+            eigs = np.linalg.eigh(X)[0]
+            eigs = eigs[eigs > 0]
```

Comment:
What is the nature of the density matrices? They must be symmetric (right?), but are they also positive semidefinite?

In any case, would it make sense to use `eigs[eigs > eps]` instead? (Where `eps` is something close to machine precision.)

Reply:
I don't think they're guaranteed to be positive semidefinite. Anyway, I ran into negative eigenvalues at times, thus that line. And zero eigenvalues were problematic; the paper defines log0=0, but numpy obviously disagrees. It's easier to drop the zero eigenvalues, which wouldn't contribute to the entropy anyway.

To make this seem marginally more principled, page 2 of the paper mentions non-negative eigenvalues specifically, and that 0log2 0 := 0.

This line was the last thing I added. The problem was that when I reproduced fig 2a of the paper, in the large-beta regime the von Neumann entropy would go to nan instead of zero.